Add support for multiple platforms in image export and load

ctalledo opened the pull request:
- needs: https://github.com/moby/moby/pull/50418

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Add support for multiple platforms when loading or exporting images from/to tar archives. 

This way clients (e.g., docker cli) can do things like:

```
docker image save --platform linux/amd64,linux/arm64/v8,riscv64 -o alpine.tar alpine:latest
```

Or

```
docker image load --platform linux/arm64/v8,riscv64 -i alpine.tar
```

Fixes #48759 .

*NOTE*: See the corresponding [PR](https://github.com/docker/cli/pull/6126) in the Docker CLI so that `docker image load` and `docker image save` accept multiple platforms via the `--platform` option (as opposed to a single platform as they currently do).

**- How I did it**

Changed the image export and load APIs so that they accept an array of platforms, as opposed to a single platform. Also modified the code that handles those APIs accordingly.

**- How to verify it**

Updated unit and integration tests accordingly.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
`GET /images/{name}/get` and `POST /images/load` now accept multiple `platform` query parameters, allowing export and load of images for multiple platforms.
```

**- A picture of a cute animal (not mandatory but encouraged)**

ctalledo committed:
Add support for multiple platforms in image export and loading.

Currently the image export and load APIs can be used to export or load all
platforms for the image, or a single specified platform.

This commit updates the API so that it accepts a list of platforms to export or
load, thereby giving clients the ability to export only selected platforms of an
image into a tar file, or load selected platforms from a tar file.

Unit and integration tests were updated accordingly.

As this requires a daemon API change, the API version was bumped.

Signed-off-by: Cesar Talledo <cesar.talledo@docker.com>

thaJeztah commented:
Looks like the description and title may not match; it's probably either `export` and `import` or `save` and `load` 😅 ?

ctalledo commented:
Hi @thaJeztah,

> Looks like the description and title may not match; it's probably either export and import or save and load 😅 ?

From a docker CLI perspective, I mean `docker save` and `docker load`. 

But the naming I used in the title reflects the naming in the Moby code itself ([exportImage](https://github.com/ctalledo/moby/blob/13879e7b496d14fb0724719c49c858731c9e7f60/daemon/containerd/image_exporter.go#L34) for `docker save` and [loadImage](https://github.com/ctalledo/moby/blob/13879e7b496d14fb0724719c49c858731c9e7f60/daemon/containerd/image_exporter.go#L232) for `docker load`.

And to make the naming even more confusing, `exportImage` corresponds to the `/images/get` API :)

So some renaming would be good here for consistency (can do that in a follow-up commit).

thaJeztah commented on daemon/images/image_exporter.go:
```diff
@@ -13,15 +13,30 @@ import (
+
+	// The docker image store only supports single platform images. If multiple platforms are given,
+	// use the last one (for compatibility with previous behavior).
+	if len(platformList) > 0 {
+		platform = platformList[len(platformList)-1]
+	}
```
Considering if we should make this check for the overlap between the given list of platforms and the image that's specified.

Basically thinking;

- `docker pull --platform linux/s390x alpine`
- I now have `alpine` on my machine, but only as `linux/s390x`
- `docker save --platform linux/s390x,linux/amd64 alpine` (save `alpine`,  but only the `s390x` and `amd64` variants, skip other variants
- ^^ we only have `s390x`, and "skip" `amd64` (which ... isn't present, so nothing to skip)

I guess that depends a bit on how we describe / consider the purpose of `--platform` in this case;

- Is it a requirement? (export if present, fail if not)
- Is it a filter? (export the given platforms, and skip other ones)

vvoland reviewed (changes requested):
Thanks, left some comment! Also needs an entry in version-history.md

ctalledo commented on api/server/router/image/backend.go:
```diff
@@ -32,9 +32,9 @@ type imageBackend interface {
 
 type importExportBackend interface {
-	LoadImage(ctx context.Context, inTar io.ReadCloser, platform *ocispec.Platform, outStream io.Writer, quiet bool) error
+	LoadImage(ctx context.Context, inTar io.ReadCloser, platformList []*ocispec.Platform, outStream io.Writer, quiet bool) error
```
Makes sense, done.

ctalledo commented on api/server/router/image/image_routes.go:
```diff
@@ -242,34 +242,38 @@ func (ir *imageRouter) getImagesGet(ctx context.Context, w http.ResponseWriter,
-		if formPlatforms := r.Form["platform"]; len(formPlatforms) > 1 {
-			// TODO(thaJeztah): remove once we support multiple platforms: see https://github.com/moby/moby/issues/48759
-			return errdefs.InvalidParameter(errors.New("multiple platform parameters not supported"))
+		formPlatforms := r.Form["platform"]
```
Thanks, done.

ctalledo commented on daemon/images/image_exporter.go:
```diff
@@ -13,15 +13,30 @@ import (
+	if len(platformList) > 0 {
+		platform = platformList[len(platformList)-1]
+	}
+
```
Agreed, done.

ctalledo commented on daemon/containerd/image_exporter.go:
```diff
@@ -31,8 +31,19 @@ import (
+			pl = append(pl, *p)
+		}
+		pm = platforms.Any(pl...)
+	}
```
Yes good catch; I've fixed it. Had to create a new matcher called `matchAnyWithPreference` which matches any of the given platforms but orders the result so that the host platform comes first (if present in the image manifests).

thaJeztah reviewed (commented):
just some quick blurbs

ctalledo commented on daemon/containerd/platform_matchers.go:
```diff
@@ -25,6 +25,28 @@ func (c allPlatformsWithPreferenceMatcher) Less(p1, p2 ocispec.Platform) bool {
+}
+
+// matchAnyWithPreference will return a platform matcher that matches any of the
+// given platforms but will order platforms matching the preferred matcher first.
```
Make sense, done.

ctalledo commented on daemon/containerd/platform_matchers.go:
```diff
@@ -25,6 +25,28 @@ func (c allPlatformsWithPreferenceMatcher) Less(p1, p2 ocispec.Platform) bool {
+
+// matchAnyWithPreference will return a platform matcher that matches any of the
+// given platforms but will order platforms matching the preferred matcher first.
+func matchAnyWithPreference(preferred platforms.MatchComparer, platformList []ocispec.Platform) platforms.MatchComparer {
```
Agree, done.

ctalledo committed:
Refactor containerd platform matchers.

Merge the matchAllWithPreference matcher into the matchAnyWithPreference
matcher to reduce code duplication and simplify the code.

Signed-off-by: Cesar Talledo <cesar.talledo@docker.com>

vvoland commented on daemon/containerd/image_save_test.go:
```diff
@@ -93,11 +111,19 @@ func TestImageMultiplatformSaveShallowWithoutNative(t *testing.T) {
+	t.Run("export multiple platforms with some missing", func(t *testing.T) {
+		err = imgSvc.ExportImage(ctx, []string{img.Name}, []ocispec.Platform{arm64, native}, io.Discard)
+		assert.Check(t, is.ErrorType(err, cerrdefs.IsNotFound))
+	})
```
Can you add a test case that tries to export platform that doesn't exist at all in the image index?

Make sure we error out instead of not saving it silently.

ctalledo replied:
Good idea, done.

vvoland reviewed (commented):
Thanks! Left some comments on edge cases.

vvoland commented on daemon/images/image_exporter.go:
```diff
@@ -4,24 +4,42 @@ import (
+	if len(platformList) > 1 {
+		return errdefs.InvalidParameter(errors.New("multiple platforms not supported for this image store; use a multi-platform image store such as containerd-snapshotter"))
+	} else if len(platformList) == 1 {
+		platform = &platformList[0]
```
No, I think we should just error out if `len(platformList) > 1` because the graphdrivers can't ever export more than one platform because all images are single platform.

ctalledo replied:
Yes agree, but that's what the code is doing already correct?

```
if len(platformList) > 1 {
		return errdefs.InvalidParameter(errors.New("multiple platforms not supported for this image store; use a multi-platform image store such as containerd-snapshotter"))
```

vvoland replied:
Oh, right indeed!

I got confused by this comment 🤦🏻 

https://github.com/moby/moby/blob/804d457488bcfda0f1599128b7381b47ce075d7d/integration/image/save_test.go#L236

ctalledo replied:
Ah good catch, sorry about that; I've removed that confusing comment.

ctalledo replied:
@vvoland: let me know if the PR looks good now, thanks!

vvoland reviewed (approved):
LGTM

thaJeztah commented on integration/image/save_test.go:
```diff
@@ -13,9 +13,9 @@ import (
-	cerrdefs "github.com/containerd/errdefs"
 	"github.com/cpuguy83/tar2go"
 	containertypes "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/image"
```
This needs a rebase, and needs to use the new import paths for the api


After rebasing and updating the import paths,  it's also needed to run hack/vendor.sh because we now vendor the API module in the main module (don't ask 😆)

thaJeztah commented:
Did a quick rebase and vendor

ctalledo commented:
> Did a quick rebase and vendor

Thanks @thaJeztah!

ctalledo commented:
CI looks good, the one integration test failure is spurious and unrelated to this PR.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -10443,7 +10443,10 @@ paths:
           required: true
         - name: "platform"
-          type: "string"
+          type: "array"
+          items:
+            type: "string"
```
I'll open a follow-up to update the docs/api/version-history.md - those docs probably still need to be moved to the API module as well.

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -266,16 +286,19 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+			for _, p := range platformList {
+				platformNames = append(platformNames, platforms.FormatAll(p))
+			}
+			log.G(ctx).WithFields(log.Fields{"error": err, "platform(s)": platformNames}).Debug("failed to import image to containerd")
```
Nit (can be updated in a follow-up); `platform(s)` as name for the structured logs feels a bit awkward; probably either `platforms` or `platform` would be better.

thaJeztah commented on daemon/containerd/image_load_test.go:
```diff
@@ -60,39 +73,64 @@ func TestImageLoadMissing(t *testing.T) {
+			err = verifyImagePlatforms(ctx, imgSvc, imgRef, platforms)
+			assert.NilError(t, err)
+		})
+		cleanup(ctx, t)
```
No need to change; I initially considered this could be a t.Cleanup` in each test; like https://go.dev/play/p/fyK3a91yn5g, but not sure if `t.Cleanup` is guaranteed to have finished between each test 🤔 (that way the cleanup - including any failure - was associated with each test).

```go
package main

import "testing"

func TestMultiCleanup(t *testing.T) {
	cleanup := func(t *testing.T) func() {
		return func() {
			t.Helper()
			t.Log("cleaning up", t.Name())
		}
	}

	t.Run("test one", func(t *testing.T) {
		t.Cleanup(cleanup(t))
	})
	t.Run("test two", func(t *testing.T) {
		t.Cleanup(cleanup(t))
		t.Fatal("test two failed")
	})
	t.Run("test three", func(t *testing.T) {
		t.Cleanup(cleanup(t))
	})
}
```

 
Alternatively,, we could  could setup a fake store as part of a `setup()` for each test, so that no state is shared between tests which would allow `t.Parallel()` to work as well.

thaJeztah commented on daemon/server/router/image/image_routes.go:
```diff
@@ -271,18 +273,18 @@ func (ir *imageRouter) postImagesLoad(ctx context.Context, w http.ResponseWriter
-			return errdefs.InvalidParameter(errors.New("multiple platform parameters not supported"))
+		var err error
+		formPlatforms := r.Form["platform"]
+		// multi-platform params were introduced in API version 1.51
+		if versions.LessThan(httputils.VersionFromContext(ctx), "1.51") && len(formPlatforms) > 1 {
+			return errdefs.InvalidParameter(errors.New("multiple platform parameters are not supported in this API version; use API version 1.51 or later."))
```
Ah, derp; this needs to be v1.52` now; https://github.com/moby/moby/blob/b0c22a931d65a56b10ba1a4e542c70ac1bd8ea1a/api/common.go#L5-L6

thaJeztah reviewed (approved):
LGTM


I can fix-up the API version and version-history in a follow-up.

thaJeztah commented:
- opened https://github.com/moby/moby/pull/50489 for fixing the API version and change-log
